WcaOnRails/app/controllers/payment_controller.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+
+class PaymentController < ApplicationController
+  def payment_config
+    competition = Competition.find(params[:competition_id])
+    render json: { stripe_publishable_key: EnvVars.STRIPE_PUBLISHABLE_KEY, connected_account_id: competition.connected_stripe_account_id }
+  end
+
+  def payment_finish
+    competition = Competition.find(params[:competition_id])
+
+    # Provided by Stripe upon redirect when the "PaymentElement" workflow is completed
+    intent_id = params[:payment_intent]
+    intent_secret = params[:payment_intent_client_secret]
+
+    stored_transaction = StripeTransaction.find_by(stripe_id: intent_id)
+    stored_intent = stored_transaction.stripe_payment_intent
+
+    unless stored_intent.client_secret == intent_secret
+      return redirect_to competition_register_path(competition, "secret_invalid")
+    end
+
+    # No need to create a new intent here. We can just query the stored intent from Stripe directly.
+    stripe_intent = stored_intent.retrieve_intent
+
+    unless stripe_intent.present?
+      return redirect_to competition_register_path(competition, "not_found")
+    end
+
+    stored_intent.update_status_and_charges(stripe_intent, current_user) do |_|
+      begin
+        update_registration_payment(stripe_intent.id, stripe_intent.status)
+      rescue Error
+        return redirect_to competition_register_path(competition_id, "Couldn't reach Registration Service")
```

Comment:
You use stringified error messages like `"not_found"` above but then suddenly a humanized English phrase here. Please be consistent in the error handling.